mlir/lib/Conversion/TosaToRock/TosaToRock.cpp

```diff
@@ -406,31 +412,89 @@ struct TransposeRewritePattern : public OpRewritePattern<tosa::TransposeOp> {
     op->setAttr(name, BoolAttr::get(op->getContext(), newValue));
   }
 
+  LogicalResult checkMatMulTransposeValid(tosa::MatMulOp matmulOp,
+                                          ArrayRef<int32_t> dims) const {
+    // batch dimension is expected to be 3rd from the last.
+    if (dims.size() >= 3 && dims[dims.size() - 3] != (int32_t)dims.size() - 3) {
+      return matmulOp.emitError(
+          "Can't transpose the batch dimension out of place");
+    }
+    return success();
+  }
+
+  bool isMatMulNonTrivial(ArrayRef<int32_t> dims) const {
+    int32_t lastDim = dims.size() - 1;
+    int32_t prevLastDim = dims.size() - 2;
+    return (dims[prevLastDim] == lastDim && dims[lastDim] == prevLastDim);
+  }
+
+  // This function traverses the uses of tOutput and then modifies
+  // the uses to indicate the input are transposed and replaces them
+  // with tInput. If there are collapse shapes encountered, the collapse
+  // is applied on the tInput.
+  LogicalResult mergeTransposeWithGemmLikeOp(PatternRewriter &rewriter,
+                                             const Value &tOutput,
```

Comment:
`Value` is a smart pointer, just pass it by value. That is, any and all uses of `const Value&` can be `Value`

Reply:
done.